### DISCUSSION TITLE
How to upgrade minio from RELEASE.2021-12-20T22-07-16Z to RELEASE.2023-04-13T03-08-07Z  ?

### DISCUSSION THREAD

**Discussion by pinpindao** [minio/minio#17056] [Q&A] [answered]
The current minio release is 2021-12-20T22-07-16Z, there are some test data,as below:
![image](https://user-images.githubusercontent.com/106652818/233323589-17c787ac-9690-4ef6-a8fb-dd8f06683f48.png)

Refer the document form https://github.com/minio/minio as below:  Upgrade failed.
![image](https://user-images.githubusercontent.com/106652818/233325589-4baf44e8-d635-4078-8379-623e65e4fb9c.png)

Steps for upgrade:
1. the current release is binary installed and systemd started.
[root@ECS-OPS-001 ~]# systemctl status minio
● minio.service - Minio service
   Loaded: loaded (/etc/systemd/system/minio.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2023-04-20 16:51:36 CST; 44min ago
     Docs: https://docs.minio.io/
 Main PID: 4225 (run.sh)
   CGroup: /system.slice/minio.service
           ├─4225 /bin/bash /data/minio/run/run.sh
           └─4226 /data/minio/run/minio server --console-address 0.0.0.0:9900 --config-dir /etc/minio /data/minio/data

Apr 20 16:51:36 ECS-OPS-001 systemd[1]: Started Minio service.
Apr 20 16:51:45 ECS-OPS-001 run.sh[4225]: WARNING: MINIO_ACCESS_KEY and MINIO_SECRET_KEY are deprecated.
Apr 20 16:51:45 ECS-OPS-001 run.sh[4225]: Please use MINIO_ROOT_USER and MINIO_ROOT_PASSWORD
Apr 20 16:51:45 ECS-OPS-001 run.sh[4225]: API: http://192.168.0.125:9000  http://127.0.0.1:9000
Apr 20 16:51:45 ECS-OPS-001 run.sh[4225]: Console: http://0.0.0.0:9900
Apr 20 16:51:45 ECS-OPS-001 run.sh[4225]: Documentation: https://docs.min.io
[root@ECS-OPS-001 ~]# cd /data/minio/
[root@ECS-OPS-001 minio]# pwd
/data/minio
[root@ECS-OPS-001 minio]# tree
.
├── data
│   └── bird
│       ├── a.txt
│       └── b.txt
├── run
│   ├── mc
│   ├── minio
│   └── run.sh
└── run.sh

3 directories, 6 files

2. Download the binary from https://dl.min.io/ and replace the existing MinIO binary
[root@ECS-OPS-001 minio]# cd run
[root@ECS-OPS-001 run]# ll
total 234104
-rwxr-xr-x 1 root root  21782528 Apr 19 17:28 mc
-rwxr-xr-x 1 root root  99561472 Apr 19 14:48 minio       --> new release
-rwxr-xr-x 1 root root 118370304 Apr 19 17:28 minio.bak
-rwxr-xr-x 1 root root       177 Apr 20 16:42 run.sh

3. Set alias:
[root@ECS-OPS-001 run]# ./mc alias set myminio http://192.168.0.125:9000 Minio 12345678
Added `myminio` successfully.
[root@ECS-OPS-001 run]# ./mc admin info myminio
●  192.168.0.125:9000
   Uptime: 50 minutes 
   Version: 2021-12-20T22:07:16Z
   Network: 1/1 OK 

68 B Used, 1 Bucket, 2 Objects

4. perform restart
[root@ECS-OPS-001 run]# ./mc admin service restart myminio
Restart command successfully sent to `myminio`. Type Ctrl-C to quit or wait to follow the status of the restart process.
..............................................................................

it is stuck in restarting status:
[root@ECS-OPS-001 ~]# systemctl status minio
● minio.service - Minio service
   Loaded: loaded (/etc/systemd/system/minio.service; disabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Thu 2023-04-20 17:50:06 CST; 1s ago
     Docs: https://docs.minio.io/
  Process: 12805 ExecStart=/data/minio/run/run.sh (code=exited, status=1/FAILURE)
 Main PID: 12805 (code=exited, status=1/FAILURE)

Apr 20 17:50:06 ECS-OPS-001 systemd[1]: Unit minio.service entered failed state.
Apr 20 17:50:06 ECS-OPS-001 systemd[1]: minio.service failed.

From syslogs,can see below messages:
Apr 20 17:46:36 ECS-OPS-001 systemd: Started Minio service.
Apr 20 17:46:36 ECS-OPS-001 run.sh: WARNING: MINIO_ACCESS_KEY and MINIO_SECRET_KEY are deprecated.
Apr 20 17:46:36 ECS-OPS-001 run.sh: Please use MINIO_ROOT_USER and MINIO_ROOT_PASSWORD
Apr 20 17:46:36 ECS-OPS-001 run.sh: WARNING: Detected Linux kernel version older than 4.0.0 release, there are some known potential performance problems with this kernel version. MinIO recommends a minimum of 4.x.x linux kernel version for best performance
Apr 20 17:46:36 ECS-OPS-001 run.sh: ERROR Unable to use the drive /data/minio/data: Drive /data/minio/data: found backend type fs, expected xl or xl-single - to migrate to a supported backend visit https://min.io/docs/minio/linux/operations/install-deploy-manage/migrate-fs-gateway.html: Invalid arguments specified
Apr 20 17:46:36 ECS-OPS-001 systemd: minio.service: main process exited, code=exited, status=1/FAILURE
Apr 20 17:46:36 ECS-OPS-001 systemd: Unit minio.service entered failed state.
Apr 20 17:46:36 ECS-OPS-001 systemd: minio.service failed.

5. When I delete the  .minio.sys directory from data, minio start successfully.
[root@ECS-OPS-001 data]# pwd
/data/minio/data
[root@ECS-OPS-001 data]# ll -a
total 16
drwxr-xr-x 4 root root 4096 Apr 20 16:51 .
drwxr-xr-x 5 root root 4096 Apr 19 18:29 ..
drwxr-xr-x 2 root root 4096 Apr 20 16:49 bird
drwxr-xr-x 7 root root 4096 Apr 20 17:49 .minio.sys
[root@ECS-OPS-001 data]# rm -rf .minio.sys
[root@ECS-OPS-001 data]# ll
total 4
drwxr-xr-x 2 root root 4096 Apr 20 16:49 bird
[root@ECS-OPS-001 data]# systemctl status minio
● minio.service - Minio service
   Loaded: loaded (/etc/systemd/system/minio.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2023-04-20 17:50:59 CST; 2min 57s ago
     Docs: https://docs.minio.io/
 Main PID: 13019 (run.sh)
   CGroup: /system.slice/minio.service
           ├─13019 /bin/bash /data/minio/run/run.sh
           └─13020 /data/minio/run/minio server --console-address 0.0.0.0:9900 --config-dir /etc/minio /data/minio/data

Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: WARNING: Host local has more than 0 drives of set. A host failure will result in data becoming unavailable.
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: MinIO Object Storage Server
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Copyright: 2015-2023 MinIO, Inc.
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Version: RELEASE.2023-04-13T03-08-07Z (go1.20.3 linux/amd64)
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Status:         1 Online, 0 Offline.
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: API: http://192.168.0.125:9000  http://127.0.0.1:9000
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Console: http://0.0.0.0:9900
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Documentation: https://min.io/docs/minio/linux/index.html
Apr 20 17:50:59 ECS-OPS-001 run.sh[13019]: Warning: The standard parity is set to 0. This can lead to data loss.

The new release minio login normally,but no data in buckets,and the bucket create time is wrong.
![image](https://user-images.githubusercontent.com/106652818/233331442-08492b14-02c8-47e4-851c-b3d015b3bd6a.png)

Try to delete bucket,it prompt not empty
![image](https://user-images.githubusercontent.com/106652818/233332554-a406b8f0-fb66-44f8-895c-45a29a5e9783.png)

So,how to upgrade the minio from the old version which have buckets,could you please share some suggestions?

**Comment by ravindk89** [comment, marked as answer]
> ERROR Unable to use the drive /data/minio/data: Drive /data/minio/data: found backend type fs, expected xl or xl-single - to migrate to a supported backend visit https://min.io/docs/minio/linux/operations/install-deploy-manage/migrate-fs-gateway.html: Invalid arguments specified

You are running in the legacy FS mode. Visit the linked doc for information on how to migrate to the new backend.

**Reply by pinpindao** [reply]
Hello,
We just want to do the release upgrade directly,but have to do migrate now,you know we have 50G+ data in the existing system.

Anyway,refer to https://min.io/docs/minio/linux/operations/install-deploy-manage/migrate-fs-gateway.html ,we try to test the migration,but have a new issue.

1. Create a new Single-Node Single-Drive MinIO deployment
minio version RELEASE.2023-03-20T20-16-18Z
mc version RELEASE.2023-04-12T02-21-51Z
Install minio on a new machine.

3. Add an alias for the new deployment with the new MinIO Client
[root@localhost ~]# mc alias set newminio http://10.0.0.157:9000 Minio 12345678
Added `newminio` successfully.

4. Export the existing deployment’s configurations
[root@CentOS7 run]# ./mc admin config export myminio > config.txt

5. Import configurations from existing standalone deployment to new deployment with the new MinIO Client
[root@localhost ~]# mc admin config import newminio < ./config.txt 
mc: <ERROR> Unable to set server config: sub-system 'heal' cannot have empty keys.

Could you please clarify it ?

![image](https://user-images.githubusercontent.com/106652818/233536449-369d0717-8e3c-4187-9f24-7954bc3eebd4.png)

**Reply by ravindk89** [reply]
You can comment out the `#heal` line - that should allow for importing the config.

@harshavardhana I think I saw this come up once before for older deployments - the general guidance seems to be just commenting out the keys in question. I'll add that to the docs, but did want to confirm I have the right idea.

**Reply by pinpindao** [reply]
I comment out the #heal line
#heal bitrotscan=off max_sleep=1s max_io=100 

But another error arisen.

[root@localhost ~]# mc admin config import newminio < config.txt
mc: <ERROR> Unable to set server config: invalid value strict replication_workers=250 replication_failed_workers=8 for list_quorum.

**Reply by ravindk89** [reply]
Maybe a better question is this - did you have any custom config settings on your original deployment? at a glance I don't see anything unusual. And replication would not have existed in your version anyways (not sure why it's in the config then).

If there's nothing custom about your setup, you can just skip importing config. It's more important for users who have IDP configurations or bucket notification settings to preserve.

**Reply by pinpindao** [reply]
My original deployment is simple,no custom config settings.
mc version RELEASE.2021-12-20T23-43-34Z
minio version RELEASE.2021-12-20T22-07-16Z
Original deployment as below, the version is consistent with the production environment, just for test now, Single-Node Single-Drive MinIO

[root@CentOS7 run]# cat /etc/systemd/system/minio.service
[Unit]
Description=Minio service
Documentation=https://docs.minio.io/
[Service]
WorkingDirectory=/data/minio/run/
ExecStart=/data/minio/run/run.sh
Restart=on-failure
RestartSec=5
[Install]
WantedBy=multi-user.target

[root@CentOS7 run]# cat /data/minio/run/run.sh
#!/bin/bash
export MINIO_ACCESS_KEY=Minio
export MINIO_SECRET_KEY=12345678
/data/minio/run/minio server --config-dir /etc/minio /data/minio/data

Skip importing config, execute step 6 --  Export bucket metadata from existing standalone deployment with the existing MinIO Client

[root@CentOS7 run]# ./mc admin cluster bucket export myminio
mc: <ERROR> `cluster` is not a recognized command. Get help using `--help` flag.

There is no `cluster` para in the original deployment version, how can I go on?

**Reply by pinpindao** [reply]
Hello,

According to  https://min.io/docs/minio/linux/operations/install-deploy-manage/migrate-fs-gateway.html 
the existing MinIO Client had no  `cluster` para, how to execute it?
![image](https://user-images.githubusercontent.com/106652818/233758467-beb8b154-6fc2-4212-92a4-c90e759eda1f.png)

Actually,my main case is how to upgrade the production environment:
4 nodes 4 drives cluster
mc version RELEASE.2021-12-20T23-43-34Z
minio version RELEASE.2021-12-20T22-07-16Z

@ravindk89 Could you please share the upgrade doc from RELEASE.2021-12-20T22-07-16Z to RELEASE.2023-04-13T03-08-07Z ?
Thanks!

**Reply by ravindk89** [reply]
I'm going to revisit the entire tutorial this week - it may need significant simplification at this point.

**Reply by ravindk89** [reply]
Though - if you are already on distributed MinIO, there's no reason to be using this tutorial. Just follow the standard upgrade procedure. test in a lower environment and then apply in production. If you only have prod, you can try 3-6 month jumps at a time, but its probably easiest to just go to latest and then rework the bits that changed.
